src/apprt/gtk-ng/class/command_palette.zig

```diff
@@ -0,0 +1,426 @@
+        self.activated(pos);
+    }
+
+    //---------------------------------------------------------------
+
+    pub fn toggle(self: *CommandPalette) void {
+        const priv = self.private();
+        const window = priv.window orelse {
+            log.warn("command palette is not associated with a window", .{});
+            return;
+        };
+
+        // Show the dialog
+        priv.dialog.present(window.as(gtk.Widget));
+        // Focus on the search bar when opening the dialog
+        _ = priv.search.as(gtk.Widget).grabFocus();
+    }
+
+    fn updateConfig(self: *CommandPalette) void {
+        const priv = self.private();
+        const config = if (priv.config) |config| config.get() else {
+            log.warn("command palette does not have a config!", .{});
+            return;
+        };
+
+        // Clear existing binds and clear allocated data
+        priv.source.removeAll();
+        _ = priv.arena.reset(.retain_capacity);
+
+        for (config.@"command-palette-entry".value.items) |command| {
+            // Filter out actions that are not implemented or don't make sense
+            // for GTK
+            switch (command.action) {
+                .close_all_windows,
+                .toggle_secure_input,
+                .check_for_updates,
+                .redo,
+                .undo,
+                .reset_window_size,
+                .toggle_window_float_on_top,
+                => continue,
+
+                else => {},
+            }
+
+            const cmd = Command.new(
+                priv.arena.allocator(),
+                command,
+                config.keybind.set,
+            ) catch |err| switch (err) {
+                error.OutOfMemory, error.NoSpaceLeft => {
+                    log.warn("unable to allocate new command err={}", .{err});
+                    return;
+                },
+            };
+            const cmd_ref = cmd.as(gobject.Object);
+            priv.source.append(cmd_ref);
+            cmd_ref.unref();
+        }
+    }
+
+    fn activated(self: *CommandPalette, pos: c_uint) void {
+        const priv = self.private();
+
+        const window = priv.window orelse {
+            log.warn("command palette is not associated with a window", .{});
+            return;
+        };
+
+        // Use priv.model and not priv.source here to use the list of *visible* results
+        const object = priv.model.as(gio.ListModel).getObject(pos) orelse return;
+        const cmd = gobject.ext.cast(Command, object) orelse return;
+
+        // Close before running the action in order to avoid being replaced by
+        // another dialog (such as the change title dialog). If that occurs then
+        // the command palette dialog won't be counted as having closed properly
+        // and cannot receive focus when reopened.
+        _ = priv.dialog.close();
+
+        const action = cmd.getAction() orelse return;
+
+        window.performBindingAction(action);
```

Comment:
Feels like this should be a signal instead and allow embedders to do what they want when an action is chosen.

Reply:
That seems like we're over-complicating things, as this is `gtk-ng` talking to `gtk-ng` so there's no libghostty embedding going on.